kimchi/src/circuits/polynomials/permutation.rs

```diff
@@ -481,27 +478,58 @@ impl<F: PrimeField, G: KimchiCurve<ScalarField = F>, OpeningProof: OpenProof<G>>
         //~ \end{align}
         //~ $$
         //~
-        for j in 0..n - 1 {
-            z[j + 1] = witness
-                .iter()
-                .zip(self.column_evaluations.permutation_coefficients8.iter())
-                .map(|(w, s)| w[j] + (s[8 * j] * beta) + gamma)
-                .fold(F::one(), |x, y| x * y);
-        }
+
+        // We compute z such that:
+        // z[0] = 1
+        // z[j+1] = \Prod_{i=0}^{COLUMNS}(wit[i][j] + (s[i][8*j] * beta) + gamma)     for j ∈ 0..n-1
```

Comment:
`COLUMNS` is not the correct variable. It is `PERMUTS`.
```
witness
            .par_iter()
            .zip(self.column_evaluations.permutation_coefficients8.par_iter())
```
will take the size of smaller iterator.

`witness` is of size `COLUMNS` (15) - the fifteen wires of Kimchi - where `[...permutation_coefficients8]` is of size `PERMUTS` (8) - the eight wires that are used by the permutation argument.